### PR TITLE
W3DTruckDraw, W3DOverlordTruckDraw and W3DPoliceCarDraw

### PR DESCRIPTION
**Pull request by jonwil**

### CODE REVIEW

File: src/game/client/draw/w3doverlordtruckdraw.cpp [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+NameKeyType W3DOverlordTruckDraw::Get_Module_Name_Key() const
+{
+    static const NameKeyType nk = g_theNameKeyGenerator->Name_To_Key("W3DOverlordTruckDraw");
+    return nk;
+}
+
+void W3DOverlordTruckDraw::Do_Draw_Module(const Matrix3D *transform)
+{
+    W3DTruckDraw::Do_Draw_Module(transform);
+    Object *object = Get_Drawable()->Get_Object();
```
Comment (xezon) [inline src/game/client/draw/w3doverlordtruckdraw.cpp:41]: `const Object`

File: src/game/client/draw/w3doverlordtruckdraw.cpp [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+void W3DOverlordTruckDraw::Load_Post_Process()
+{
+    W3DTruckDraw::Load_Post_Process();
+}
+
+void W3DOverlordTruckDraw::Set_Hidden(bool hidden)
+{
+    W3DTruckDraw::Set_Hidden(hidden);
+    Object *object = Get_Drawable()->Get_Object();
```
Comment (xezon) [inline src/game/client/draw/w3doverlordtruckdraw.cpp:77]: `const Object`

File: src/game/client/draw/w3doverlordtruckdraw.h [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+};
+
+class W3DOverlordTruckDraw : public W3DTruckDraw
+{
+    IMPLEMENT_POOL(W3DOverlordTruckDraw);
+
+public:
+    W3DOverlordTruckDraw(Thing *thing, ModuleData const *module_data) : W3DTruckDraw(thing, module_data) {}
+
+    virtual ~W3DOverlordTruckDraw() override {}
```
Comment (xezon) [inline src/game/client/draw/w3doverlordtruckdraw.h:43]: `protected`

File: src/game/client/draw/w3dpolicecardraw.cpp [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+}
+
+void W3DPoliceCarDraw::CRC_Snapshot(Xfer *xfer)
+{
+    W3DTruckDraw::CRC_Snapshot(xfer);
+}
+
+void W3DPoliceCarDraw::Xfer_Snapshot(Xfer *xfer)
+{
+    unsigned char ver = 1;
```
Comment (xezon) [inline src/game/client/draw/w3dpolicecardraw.cpp]: `uint8_t`

File: src/game/client/draw/w3dpolicecardraw.h [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+class W3DDynamicLight;
+
+class W3DPoliceCarDraw : public W3DTruckDraw
+{
+    IMPLEMENT_POOL(W3DPoliceCarDraw);
+
+public:
+    W3DPoliceCarDraw(Thing *thing, ModuleData const *module_data);
+
+    virtual ~W3DPoliceCarDraw() override;
```
Comment (xezon) [inline src/game/client/draw/w3dpolicecardraw.h:29]: `protected`

File: src/game/client/draw/w3dtruckdraw.h [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+ *            LICENSE
+ */
+#pragma once
+#include "always.h"
+#include "w3dmodeldraw.h"
+
+class W3DTruckDrawModuleData : public W3DModelDrawModuleData
+{
+public:
+    W3DTruckDrawModuleData() {}
```
Comment (xezon) [inline src/game/client/draw/w3dtruckdraw.h]: Constructor does not initialize anything?

File: src/game/client/draw/w3dtruckdraw.cpp [hunk 7]
```diff
@@ -0,0 +1,631 @@
...
+    m_midFrontLeftTireBone(0),
+    m_midFrontRightTireBone(0),
+    m_midRearLeftTireBone(0),
+    m_midRearRightTireBone(0),
+    m_midMidLeftTireBone(0),
+    m_midMidRightTireBone(0),
+    m_prevRenderObj(nullptr)
+{
+    m_landingSound = *thing->Get_Template()->Get_Per_Unit_Sound("TruckLandingSound");
+    m_powerslideSound = *thing->Get_Template()->Get_Per_Unit_Sound("TruckPowerslideSound");
```
Comment (xezon) [inline src/game/client/draw/w3dtruckdraw.cpp:93]: Not initialized members:
```c++
    int m_cabBone;
    float m_cabRotation;
    int m_trailerBone;
    float m_trailerRotation;
    int m_numBones;
```

File: src/game/client/draw/w3dtruckdraw.cpp [hunk 8]
```diff
@@ -0,0 +1,631 @@
...
+}
+
+void W3DTruckDraw::CRC_Snapshot(Xfer *xfer)
+{
+    W3DModelDraw::CRC_Snapshot(xfer);
+}
+
+void W3DTruckDraw::Xfer_Snapshot(Xfer *xfer)
+{
+    unsigned char ver = 1;
```
Comment (xezon) [inline src/game/client/draw/w3dtruckdraw.cpp]: `uint8_t`

File: src/game/client/draw/w3dtruckdraw.cpp [hunk 9]
```diff
@@ -0,0 +1,631 @@
...
+                Object *object = Get_Drawable()->Get_Object();
+
+                if (object != nullptr) {
+                    if (Get_Render_Object() != nullptr) {
+                        if (Get_Render_Object() != m_prevRenderObj) {
+                            captainslog_debug("W3DTruckDraw::Do_Draw_Module - shouldn't update bones.");
+                            Update_Bones();
+                        }
+
+                        PhysicsBehavior *physics = object->Get_Physics();
```
Comment (xezon) [inline src/game/client/draw/w3dtruckdraw.cpp:128]: `const PhysicsBehavior *`

File: src/game/client/draw/w3dtruckdraw.cpp [hunk 10]
```diff
@@ -0,0 +1,631 @@
...
+
+                        if (physics != nullptr) {
+                            const Coord3D &velocity = physics->Get_Velocity();
+                            float magnitude = physics->Get_Velocity_Magnitude();
+                            TWheelInfo *info = Get_Drawable()->Get_Wheel_Info();
+                            AIUpdateInterface *update = object->Get_AI_Update_Interface();
```
Comment (xezon) [inline src/game/client/draw/w3dtruckdraw.cpp:134]: `const AIUpdateInterface *`

File: src/game/client/draw/w3dtruckdraw.cpp [hunk 11]
```diff
@@ -0,0 +1,631 @@
...
+                            captainslog_debug("W3DTruckDraw::Do_Draw_Module - shouldn't update bones.");
+                            Update_Bones();
+                        }
+
+                        PhysicsBehavior *physics = object->Get_Physics();
+
+                        if (physics != nullptr) {
+                            const Coord3D &velocity = physics->Get_Velocity();
+                            float magnitude = physics->Get_Velocity_Magnitude();
+                            TWheelInfo *info = Get_Drawable()->Get_Wheel_Info();
```
Comment (xezon) [inline src/game/client/draw/w3dtruckdraw.cpp]: `const TWheelInfo *info`

I think `TWheelInfo *Get_Wheel_Info() const` can also return `const TWheelInfo *`

File: src/game/client/draw/w3dtruckdraw.cpp [hunk 12]
```diff
@@ -0,0 +1,631 @@
...
+
+    if (g_theWriteableGlobalData->m_showClientPhysics) {
+        const W3DTruckDrawModuleData *data = Get_W3D_Truck_Draw_Module_Data();
+
+        if (data != nullptr) {
+            bool frozen = g_theTacticalView->Is_Time_Frozen() && !g_theTacticalView->Is_Camera_Movement_Finished();
+            frozen = frozen || g_theScriptEngine->Is_Time_Frozen_Debug() || g_theScriptEngine->Is_Time_Frozen_Script();
+
+            if (!frozen) {
+                Object *object = Get_Drawable()->Get_Object();
```
Comment (xezon) [inline src/game/client/draw/w3dtruckdraw.cpp:119]: `const Object *`

File: src/game/client/draw/w3dtruckdraw.cpp [hunk 13]
```diff
@@ -0,0 +1,631 @@
...
+
+                        if (physics != nullptr) {
+                            const Coord3D &velocity = physics->Get_Velocity();
+                            float magnitude = physics->Get_Velocity_Magnitude();
+                            TWheelInfo *info = Get_Drawable()->Get_Wheel_Info();
+                            AIUpdateInterface *update = object->Get_AI_Update_Interface();
+
+                            if (info != nullptr && m_cabBone != 0) {
+                                Matrix3D m(true);
+                                m.Make_Identity();
```
Comment (xezon) [inline src/game/client/draw/w3dtruckdraw.cpp:138]: Matrix3D is set to identity twice here.

File: src/game/client/draw/w3dtruckdraw.cpp [hunk 14]
```diff
@@ -0,0 +1,631 @@
...
+                                float angle = info->m_wheelAngle * data->m_cabRotationMultiplier;
+
+                                if (update != nullptr && update->Get_Path() != nullptr) {
+                                    Coord3D pos;
+                                    update->Get_Path()->Get_Point_Pos(&pos);
+                                    float angle2d = g_thePartitionManager->Get_Relative_Angle_2D(object, &pos);
+
+                                    if (angle2d >= 0.0f) {
+                                        if (angle > angle2d) {
```
Comment (xezon) [inline src/game/client/draw/w3dtruckdraw.cpp]: Can be simplified as `std::clamp(0.0f, angle2d)`

File: src/game/client/draw/w3dtruckdraw.cpp [hunk 15]
```diff
@@ -0,0 +1,631 @@
...
+                                            angle = angle2d;
+                                        }
+
+                                        if (angle < 0.0f) {
+                                            angle = 0.0f;
+                                        }
+                                    } else {
+                                        if (angle < angle2d) {
```
Comment (xezon) [inline src/game/client/draw/w3dtruckdraw.cpp]: Can be simplified as `std::clamp(angle2d, 0.0f)`

File: src/game/client/draw/w3dtruckdraw.cpp [hunk 16]
```diff
@@ -0,0 +1,631 @@
...
+
+                                    m.Make_Identity();
+                                    m.Rotate_Y(m_rearWheelRotation);
+                                    m.Adjust_Z_Translation(info->m_rearRightHeightOffset);
+                                    Get_Render_Object()->Capture_Bone(m_rearRightTireBone);
+                                    Get_Render_Object()->Control_Bone(m_rearRightTireBone, m);
+                                }
+
+                                if (m_midFrontLeftTireBone != 0 && info) {
+                                    m.Adjust_Z_Translation(info->m_frontLeftHeightOffset);
```
Comment (xezon) [inline src/game/client/draw/w3dtruckdraw.cpp:240]: `m.Make_Identity();` missing here?

File: src/game/client/draw/w3dtruckdraw.cpp [hunk 17]
```diff
@@ -0,0 +1,631 @@
...
+                                    m_rearWheelRotation += rotation * magnitude;
+                                }
+
+                                m_midFrontWheelRotation = m_frontWheelRotation;
+                                m_midRearWheelRotation = m_rearWheelRotation;
+
+                                Matrix3D m(true);
+
+                                if (m_frontLeftTireBone != 0 && info) {
+                                    m.Adjust_Z_Translation(info->m_frontLeftHeightOffset);
```
Comment (xezon) [inline src/game/client/draw/w3dtruckdraw.cpp:210]: Maybe add `m.Make_Identity();` here and remove the `true` from `Matrix3D m(true); above`

File: src/game/client/draw/w3dtruckdraw.cpp [hunk 18]
```diff
@@ -0,0 +1,631 @@
...
+                                } else {
+                                    m_rearWheelRotation += rotation * magnitude;
+                                }
+
+                                m_midFrontWheelRotation = m_frontWheelRotation;
+                                m_midRearWheelRotation = m_rearWheelRotation;
+
+                                Matrix3D m(true);
+
+                                if (m_frontLeftTireBone != 0 && info) {
```
Comment (xezon) [inline src/game/client/draw/w3dtruckdraw.cpp]: Maybe put all conditions into one `if (info)` ?

File: src/game/client/draw/w3dtruckdraw.cpp [hunk 19]
```diff
@@ -0,0 +1,631 @@
...
+
+                            if (physics->Is_Motive() && !object->Is_Significantly_Above_Terrain()) {
+                                Enable_Emitters(true);
+                                Coord3D prev = physics->Get_Prev_Accel();
+                                Coord3D c;
+                                c.x = prev.x;
+                                c.y = prev.y;
+                                c.z = 0.0f;
+                                float len = c.Length();
+                                bool b = len > 0.01f;
```
Comment (xezon) [inline src/game/client/draw/w3dtruckdraw.cpp]: `is_moving`

File: src/game/client/draw/w3dtruckdraw.cpp [hunk 20]
```diff
@@ -0,0 +1,631 @@
...
+            Create_Emitters();
+        }
+    }
+
+    W3DModelDraw::Set_Fully_Obscured_By_Shroud(obscured);
+}
+
+void W3DTruckDraw::On_Render_Obj_Recreated()
+{
+    m_prevRenderObj = 0;
```
Comment (xezon) [inline src/game/client/draw/w3dtruckdraw.cpp]: `nullptr`

File: src/game/client/draw/w3doverlordtruckdraw.cpp (hunk 1 above)
Comment (jonwil) [reply src/game/client/draw/w3doverlordtruckdraw.cpp:41]: Can confirm that this is calling the non-const version of Get_Object()

File: src/game/client/draw/w3doverlordtruckdraw.cpp (hunk 2 above)
Comment (jonwil) [reply src/game/client/draw/w3doverlordtruckdraw.cpp:77]: Can confirm that this is calling the non-const version of Get_Object()

File: src/game/client/draw/w3dtruckdraw.h (hunk 6 above)
Comment (jonwil) [reply src/game/client/draw/w3dtruckdraw.h]: Fixed to init all members.

File: src/game/client/draw/w3dtruckdraw.cpp (hunk 20 above)
Comment (jonwil) [reply src/game/client/draw/w3dtruckdraw.cpp]: fixed.

File: src/game/client/draw/w3dtruckdraw.cpp (hunk 19 above)
Comment (jonwil) [reply src/game/client/draw/w3dtruckdraw.cpp]: fixed.

File: src/game/client/draw/w3doverlordtruckdraw.h (hunk 3 above)
Comment (jonwil) [reply src/game/client/draw/w3doverlordtruckdraw.h:43]: fixed.

File: src/game/client/draw/w3dpolicecardraw.cpp (hunk 4 above)
Comment (jonwil) [reply src/game/client/draw/w3dpolicecardraw.cpp]: fixed.

File: src/game/client/draw/w3dpolicecardraw.h (hunk 5 above)
Comment (jonwil) [reply src/game/client/draw/w3dpolicecardraw.h:29]: fixed.

File: src/game/client/draw/w3dtruckdraw.cpp (hunk 7 above)
Comment (jonwil) [reply src/game/client/draw/w3dtruckdraw.cpp:93]: fixed.

File: src/game/client/draw/w3dtruckdraw.cpp (hunk 8 above)
Comment (jonwil) [reply src/game/client/draw/w3dtruckdraw.cpp]: fixed.

File: src/game/client/draw/w3dtruckdraw.cpp (hunk 9 above)
Comment (jonwil) [reply src/game/client/draw/w3dtruckdraw.cpp:128]: Can confirm that this is calling the non-const version of Get_Physics()

File: src/game/client/draw/w3dtruckdraw.cpp (hunk 11 above)
Comment (jonwil) [reply src/game/client/draw/w3dtruckdraw.cpp]: fixed.

File: src/game/client/draw/w3dtruckdraw.cpp (hunk 12 above)
Comment (jonwil) [reply src/game/client/draw/w3dtruckdraw.cpp:119]: Can confirm that this is calling the non-const version of Get_Object()

File: src/game/client/draw/w3dtruckdraw.cpp (hunk 16 above)
Comment (jonwil) [reply src/game/client/draw/w3dtruckdraw.cpp:240]: fixed.

File: src/game/client/draw/w3dtruckdraw.cpp (hunk 17 above)
Comment (jonwil) [reply src/game/client/draw/w3dtruckdraw.cpp:210]: fixed.

File: src/game/client/draw/w3dtruckdraw.cpp (hunk 13 above)
Comment (jonwil) [reply src/game/client/draw/w3dtruckdraw.cpp:138]: fixed.

File: src/game/client/draw/w3dtruckdraw.cpp (hunk 10 above)
Comment (jonwil) [reply src/game/client/draw/w3dtruckdraw.cpp:134]: Can confirm that this is calling the non-const version of Get_AI_Update_Interface()

**Comment by xezon** [comment]
I will test later this day.

**Review by xezon — APPROVED** [review]
Cars look as usual in game. Check with @tomsons26 & co if anything else needs to be looked at.

File: src/game/client/draw/w3dtruckdraw.cpp [hunk 21]
```diff
@@ -0,0 +1,639 @@
...
+void W3DTruckDraw::Create_Emitters()
+{
+    const W3DTruckDrawModuleData *data = Get_W3D_Truck_Draw_Module_Data();
+    if (!Get_Drawable()->Is_Hidden() && data != nullptr) {
+        if (m_dustEffect == nullptr) {
+            ParticleSystemTemplate *tmplate = g_theParticleSystemManager->Find_Template(data->m_dustEffectName);
+
+            if (tmplate != nullptr) {
+                m_dustEffect = g_theParticleSystemManager->Create_Particle_System(tmplate, true);
+                m_dustEffect->Attach_To_Drawable(Get_Drawable());
```
Comment (tomsons26) [inline src/game/client/draw/w3dtruckdraw.cpp]: this seems to be attach to object tho?

Comment (jonwil) [reply src/game/client/draw/w3dtruckdraw.cpp]: Fixed.

**Review by tomsons26 — APPROVED** [review]

**Pull request merged**